AXI4 fragmenter

terpstra opened the pull request:
This AXI4 fragmenter slices all AXI accesses into simple power-of-two sized and aligned transfers of the largest size supported by the slave.  This makes it suitable as a first stage transformation to apply before an AXI4=>TL bridge. It also makes it suitable for placing after TL=>AXI4 bridge driving an AXI-lite slave.

terpstra commented:
Rebased the PR against current master.

terpstra commented:
@henrystyles We should try this out on the VC707.

terpstra commented:
@hcook Any chance of a quick code review?

hcook commented on src/main/scala/uncore/axi4/Fragmenter.scala:
```diff
@@ -0,0 +1,268 @@
+      }
+
+      val last = beats1 === len
+      a.ready := out.ready && last
```
Correction - AXI is and stay irrevocable, so this is fine.

hcook commented on src/main/scala/uncore/axi4/Fragmenter.scala:
```diff
@@ -0,0 +1,268 @@
+    in_b.bits := out_b.bits
+
+    // Merge errors from dropped B responses
+    val r_resp = Reg(UInt(width = AXI4Parameters.respBits))
```
initialize to 0

hcook commented on src/main/scala/uncore/axi4/Test.scala:
```diff
@@ -0,0 +1,61 @@
+class RRTest0(address: BigInt) extends AXI4RegisterRouter(address, 0, 32, 0, 4)(
+  new AXI4RegBundle((), _)    with RRTest0Bundle)(
+  new AXI4RegModule((), _, _) with RRTest0Module)
+
```
nice reuse

terpstra commented on src/main/scala/uncore/axi4/Fragmenter.scala:
```diff
@@ -0,0 +1,268 @@
+      // Things that cause us to degenerate to a single beat
+      val fixed = a.bits.burst === AXI4Parameters.BURST_FIXED
+      val narrow = a.bits.size =/= UInt(lgBytes)
+      val misaligned = lo =/= UInt(0)
```
Not only can it be handled this way, but on reflection, it must be handled this way.
Otherwise, the fragmenter can pass through a misaligned access, which will cause a TL2 Monitor assertion failure.

terpstra commented:
@hcook I've addressed your concerns. However, I had to modify the sideband state machine quite a bit. Please take a peak if you could. I'm running some stress tests against it.
